### PR TITLE
8292202: modules_do is called without Module_lock

### PR DESCRIPTION
**Pull request by calvinccheung**
Please review this fix which involves:

- acquire the `Module_lock` before calling `modules_do`;
- move the call to `ClassLoader::setup_module_search_path `out of `do_module`;
- save the module paths in a `GrowableArray` in `do_module` for calling `ClassLoader::setup_module_search_path`.

Passed tiers 1 - 3 testing.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292202](https://bugs.openjdk.org/browse/JDK-8292202): modules_do is called without Module_lock

### Reviewers
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to [46fc54a0](https://git.openjdk.org/jdk/pull/10291/files/46fc54a0f3a4c3131dde27a2000bdf37861332be)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [9f4e67f9](https://git.openjdk.org/jdk/pull/10291/files/9f4e67f901d3b9ff068399099e830ec0ec8bf1e6)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10291/head:pull/10291` \
`$ git checkout pull/10291`

Update a local copy of the PR: \
`$ git checkout pull/10291` \
`$ git pull https://git.openjdk.org/jdk pull/10291/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10291`

View PR using the GUI difftool: \
`$ git pr show -t 10291`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10291.diff">https://git.openjdk.org/jdk/pull/10291.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/classfile/classLoaderExt.cpp [hunk 1]
```diff
@@ -81,20 +81,35 @@ void ClassLoaderExt::setup_app_search_path(JavaThread* current) {
...
+    ModulePathsGatherer(JavaThread* current, GrowableArray<char*>* module_paths) :
+      _current(current), _total_count(0), _module_paths(module_paths) {}
     void do_module(ModuleEntry* m) {
       char* path = m->location()->as_C_string();
       if (strncmp(path, "file:", 5) == 0) {
         path = ClassLoader::skip_uri_protocol(path);
-        ClassLoader::setup_module_search_path(_current, path);
+        char* path_copy = NEW_RESOURCE_ARRAY(char, strlen(path) + 1);
+        strcpy(path_copy, path);
+        _module_paths->at_put_grow(_total_count++, path_copy);
```
Comment (iklam) [inline src/hotspot/share/classfile/classLoaderExt.cpp:99]: You can use `_module_paths->append()` to avoid keeping the `_total_count`

**Review by iklam — APPROVED** [review]
Looks good. A minor suggestion to simplify the code.

File: src/hotspot/share/classfile/classLoaderExt.cpp (hunk 1 above)
Comment (calvinccheung) [reply src/hotspot/share/classfile/classLoaderExt.cpp]: Fixed.
Thanks for your review.

**Review by coleenp — APPROVED** [review]
Looks great.  Thanks for fixing this.  Can you add in moduleEntry.cpp in both modules_do function an assert that the Module_lock is held.  The other modules_do has the assert at a higher level so nothing will trigger it.  The modules_do() that this calls now will not fail the assert.

**Comment by calvinccheung** [comment]
Hi Coleen,

Thanks for your review.

I've added assert in moduleEntry.cpp as follows:

```
diff --git a/src/hotspot/share/classfile/moduleEntry.cpp b/src/hotspot/share/classfile/moduleEntry.cpp
index a85c8f41145..f12f63ba2e5 100644
--- a/src/hotspot/share/classfile/moduleEntry.cpp
+++ b/src/hotspot/share/classfile/moduleEntry.cpp
@@ -694,6 +694,7 @@ void ModuleEntryTable::modules_do(void f(ModuleEntry*)) {
   auto do_f = [&] (const SymbolHandle& key, ModuleEntry*& entry) {
     f(entry);
   };
+  assert_lock_strong(Module_lock);
   _table.iterate_all(do_f);
 }
 
@@ -701,6 +702,7 @@ void ModuleEntryTable::modules_do(ModuleClosure* closure) {
   auto do_f = [&] (const SymbolHandle& key, ModuleEntry*& entry) {
     closure->do_module(entry);
   };
+  assert_lock_strong(Module_lock);
   _table.iterate_all(do_f);
 }
```
I'll run some mach5 testing before integrating the fix.

**Review by coleenp — APPROVED** [review]
thanks!

**Comment by calvinccheung** [comment]
Thanks @iklam and @coleenp.

/integrate

**Pull request closed without merging**